### PR TITLE
optimize a bert model converted using tf2onnx

### PR DESCRIPTION
**Pull request by wangyems**
this PR supports some fusions for Tracy's model.

### CODE REVIEW

File: onnxruntime/python/tools/transformers/onnx_model_bert_tf.py [hunk 1]
```diff
@@ -315,4 +421,5 @@ def remove_reshape_before_first_attention(self):
 
     def postprocess(self):
         self.remove_reshape_before_first_attention()
-        self.prune_graph()
+        # Temporary work around for the following comment as it will cause topological issues for a bert model
+        #self.prune_graph()
```
Comment (wangyems) [inline onnxruntime/python/tools/transformers/onnx_model_bert_tf.py:425]: I think that prune_graph() doesn't do anything wrong here, it deletes an unnecessary bunch of nodes. It's weird that it throws ORT error for this model by reporting a graph that's not existed any more. We probably want to accommodate for this model

File: onnxruntime/python/tools/transformers/test_optimizer.py [hunk 2]
```diff
@@ -30,6 +30,7 @@
     "gpt2_past": ('gpt2_pytorch1.5_opset11', 'gpt2_past.onnx'),
     "gpt2_past_mask": ('FUSION', 'gpt2_past_mask_one_layer.onnx'),
     "multiple_embed": ('FUSION', 'embed_layer_norm_multiple.onnx'),
+    "bert_tf2onnx_0": ('other_models', 'bert_tf2onnx_0.onnx')
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/test_optimizer.py:33]: The test model shall be added to PR.

File: onnxruntime/python/tools/transformers/onnx_model_bert_tf.py [hunk 3]
```diff
@@ -295,6 +295,112 @@ def process_embedding(self):
...
+            if not self.has_constant_input(mask_nodes[1], 1):
+                logger.debug("Sub node expected to have an input with constant value 1.0.")
+                continue
+
+            upper_mask_nodes = self.match_parent_path(mask_nodes[2], ['Reshape', 'Cast', 'Concat'], [0, 1, 0])
+            if upper_mask_nodes is None:
+                continue
+            mask_concat = upper_mask_nodes[2]
+            if len(mask_concat.input) == 3:
+                # Temporary work around[Tracy's model]: require 2-d mask input, the current model has a 3-d mask input
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/onnx_model_bert_tf.py:363]: TODO: adding support of 3D mask in attention op.

File: onnxruntime/python/tools/transformers/onnx_model_bert_tf.py [hunk 4]
```diff
@@ -295,6 +295,112 @@ def process_embedding(self):
...
+
+            is_same_root = self.check_attention_input(matmul_q, matmul_k, matmul_v, parent, output_name_to_node)
+            if is_same_root:
+                mask_index = self.attention_mask.process_mask(mask_nodes[-1].input[0])
+                logger.debug("Create an Attention node.")
+                attention_node = self.attention_fusion.create_attention_node(mask_index, matmul_q, matmul_k, matmul_v,
+                                                                             add_q, add_k, add_v, parent.output[0],
+                                                                             reshape_qkv.output[0])
+                if parent.op_type == 'Reshape':
+                    # Temporary work around[Tracy's model]: we require the skiplayernorm and attention op be fed with 3-d input
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/onnx_model_bert_tf.py:378]: If the input to reshape node is 3d, we shall use that as input.

File: onnxruntime/python/tools/transformers/onnx_model_bert_tf.py [hunk 5]
```diff
@@ -295,6 +295,112 @@ def process_embedding(self):
...
+                    hidden_size = numpy_helper.to_array(self.get_initializer(parent.input[1]))[1]
+                    tensor = self.convert_list_to_tensor(parent.name + "_modified", TensorProto.INT64, [3],
+                                                         [1, -1, hidden_size])
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/onnx_model_bert_tf.py:381]: Why hard code batch_size to 1? That might not work for model with batch size > 1.

File: onnxruntime/python/tools/transformers/test_optimizer.py (hunk 2 above)
Comment (wangyems) [reply onnxruntime/python/tools/transformers/test_optimizer.py:33]: uploaded

File: onnxruntime/python/tools/transformers/onnx_model_bert_tf.py (hunk 4 above)
Comment (wangyems) [reply onnxruntime/python/tools/transformers/onnx_model_bert_tf.py:378]: for this model, the input dimension is 2-d

File: onnxruntime/python/tools/transformers/onnx_model_bert_tf.py (hunk 5 above)
Comment (wangyems) [reply onnxruntime/python/tools/transformers/onnx_model_bert_tf.py:381]: this model looks like not support batches, as the input shape is 2-d. We could probably add TODO for supporting both 3D mask and 2D input in attention node later

File: onnxruntime/python/tools/transformers/onnx_model_bert_tf.py (hunk 3 above)
Comment (wangyems) [reply onnxruntime/python/tools/transformers/onnx_model_bert_tf.py:363]: ok. Will add in this PR

File: onnxruntime/python/tools/transformers/fusion_layernorm.py [hunk 6]
```diff
@@ -121,24 +121,28 @@ def __init__(self, model: OnnxModel):
...
-         |          |                                                                            |       ^              ^
-         |          |                                                                            |       |              |
-         |          +----------------------------------------------------------------------------|-------+              |
-         |                                                                                       v                      |
-         +-------------------------------------------------------------------------------------> Mul--------------------+
+        Layer Norm from Tensorflow model(using keras2onnx or tf2onnx):
+         +------------------------------------+
+         |                                    |
+         |                                    |
+       (Cast_1)                               |
```
Comment (tianleiwu) [inline onnxruntime/python/tools/transformers/fusion_layernorm.py:128]: Cast will be removed by CastElimination (optimizer will use ORT level 1 optimization first). I think there is no need to change this function.

Comment (wangyems) [reply onnxruntime/python/tools/transformers/fusion_layernorm.py:128]: Ah I see

Comment (wangyems) [reply onnxruntime/python/tools/transformers/fusion_layernorm.py:128]: actually I found that ORT level 1 optimization does not remove those cast node

File: onnxruntime/contrib_ops/cpu/bert/attention.cc [hunk 7]
```diff
@@ -59,7 +59,8 @@ Status AttentionBase::CheckInputs(const TensorShape& input_shape,
   //   input       : (batch_size, sequence_length, hidden_size)
   //   weights     : (hidden_size, 3 * hidden_size)
   //   bias        : (3 * hidden_size)
-  //   mask_index  : nullptr, (batch_size), (2 * batch_size), (batch_size, 1), (1, 1) or (batch_size, past_sequence_length + sequence_length)
+  //   mask_index  : nullptr, (batch_size), (2 * batch_size), (batch_size, 1), (1, 1), (batch_size, past_sequence_length + sequence_length)
+  //                 or (batch_size, 1, past_sequence_length + sequence_length)
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cpu/bert/attention.cc:63]: Since the final mask data is (B, N, S, S*), so 3D input mask is better to be (B, S, S*) or (batch_size, sequence_length, past_sequence_length+sequence_length).

For (B, 1, S*), we can add a squeeze node to convert it to (B, S*) in fusion, and no need to change the operator.

Comment (wangyems) [reply onnxruntime/contrib_ops/cpu/bert/attention.cc]: added

**Comment by wangyems** [comment]
blocking for optimizing more bert models later

**Comment by hanbitmyths** [comment]
Can you add perf number from this PR compared to tf->pytorch->onnx?

**Comment by wangyems** [comment]
> Can you add perf number from this PR compared to tf->pytorch->onnx?

Note that the route  tf->pytorch->onnx may only apply to limited cases, so if it works I'll add perf number of them

**Review by hanbitmyths — APPROVED** [review]
Workaround codes needs to be maintained
